### DISCUSSION TITLE
Deploying Python Serverless Functions to Vercel with GitHub Actions

### DISCUSSION THREAD

**Discussion by snkv04** [vercel/community#7092] [Help] [answered]
### Summary

I'd like to use GitHub actions to deploy my Python runtime functions to production.

Here's my file structure:
![image](https://github.com/vercel/community/assets/45131598/91a0c6b1-2d8e-4015-a105-bc7aefdc722a)

Here is my GitHub action script for production:
https://gist.github.com/navincodesalot/b74af265eec87f67491da04285f95347

Here's the error that we're coming across:
![image](https://github.com/vercel/community/assets/45131598/fa588394-401b-4b3d-983e-7fe580f54190)

It's able to build properly, but the problem is deploying. Also, I don't need Node installed in my GitHub action in order to run the Vercel CLI for my Python functions, right?

### Steps to Reproduce

Images and logs of current project are given in summary above

**Comment by Aryan95614** [comment]
I am having the same issue, hopefully someone comes out with the solution.

**Comment by leerob** [comment]
Could you start by creating a reproducible example?

1. Start by removing things like GitHub Actions or unnecessary files
2. Limit to a single API endpoint

https://vercel.com/guides/creating-a-minimal-reproducible-example

**Reply by navincodesalot** [reply]
Hey, we seemed to have fixed this. We don't build on GitHub actions anymore, and directly build onto Vercel (our minutes are needed for our NextJS frontend app). So we don't upload a prebuilt cache of our app to Vercel. We were just wondering if there were any ways to speed up our build times as it takes the longest to install our Python dependencies. Is there a way to build the dependencies once in Vercel and then draw in those dependencies on every deployment? That'd be wonderful.

Updated github actions for reference:
https://gist.github.com/navincodesalot/e672c9448adf04ac5266834e9cd19996

**Reply by janory** [reply, marked as answer]
Hi @navincodesalot. 👋 

> Is there a way to build the dependencies once in Vercel and then draw in those dependencies on every deployment?

I think this should be possible by using Vercel's build cache.
Can you please try to set the environment variable `PIP_CACHE_DIR` to `.vercel/cache`?

**Reply by navincodesalot** [reply]
Thanks for responding. Do we set this in the github actions yaml file?

`env:
PIP_CACHE_DIR : .vercel/cache`

**Reply by navincodesalot** [reply]
> Hi @navincodesalot. 👋
> 
> > Is there a way to build the dependencies once in Vercel and then draw in those dependencies on every deployment?
> 
> I think this should be possible by using Vercel's build cache. Can you please try to set the environment variable `PIP_CACHE_DIR` to `.vercel/cache`?

Hey are there docs on this? also for nextjs and npm cache

**Comment by esebastian** [comment]
I'm struggling with the same original issue.

In my case, I need to build on GitHub Actions, and I keep getting the `Error: ENOENT: no such file or directory, lstat '/vercel/path0/.vercel/output/builds.json'` error message no matter what I try (different dependencies, CLI options, CLI versions, runtime versions...).

I've prepared a minimal reproducible example in https://github.com/esebastian/vercel-gha-flask-hello-world. The error can be seen in the workflow run logs at https://github.com/esebastian/vercel-gha-flask-hello-world/actions/runs/10167144795/job/28118854330.

**Reply by navincodesalot** [reply]
Hey so this is our workflow:
https://gist.github.com/navincodesalot/cd5aeb9bae7a4fe3f2d2f5dd3aea4951

We didn't build, we directly pushed to vercel after setting up python

**Reply by esebastian** [reply]
Thanks @navincodesalot, I actually took a look at it!

However, I need to build on GitHub Actions so I can do some tree-shaking afterwards and reduce the weight of our dependencies (otherwise Vercel complains about the size of the deployment)

The build phase goes well, but the `vercel deploy --prebuilt` always fails with `Error: ENOENT: no such file or directory, lstat '/vercel/path0/.vercel/output/builds.json'` (tree-shaking or not) as @snkv04 and @Aryan95614 reported

Maybe @leerob or @janory can provide some additional insights?

[a comment by an automated account is not shown]
